### PR TITLE
Tracer flow rate at boundaries

### PR DESCRIPTION
**Pull request by oguevremont**
### Description

An application of the tracer physics is to obtain the residence time distribution (RTD) of a flow through a domain. This PR adds the post-processing option that outputs the tracer flow rate at every boundary, providing the data for RTD analyses.

In addition, a bug where the conditions to allow `boundary conditions` to update their values were too strict was rectified. Each physics is now responsible for deciding when to update their boundary conditions, and this condition was already based on whether boundary conditions were time-dependent. 

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [ ] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If the fix is temporary, an issue is opened
- [ ] The PR description is cleaned and ready for merge

### CODE REVIEW

File: include/solvers/tracer.h [hunk 1]
```diff
@@ -386,9 +401,11 @@ class Tracer : public AuxiliaryPhysics<dim, GlobalVectorType>
 
   // Finite element spce
```
Comment (AmishgaAlphonius) [inline include/solvers/tracer.h:402]:
```suggestion
  // Finite element space
```

File: applications_tests/lethe-fluid/tracer_flow_rate.prm [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+end
+
+subsection linear solver
+  subsection fluid dynamics
+    set verbosity = quiet
+  end
+  subsection tracer
+    set verbosity = quiet
+  end
+end
```
Comment (AmishgaAlphonius) [inline applications_tests/lethe-fluid/tracer_flow_rate.prm:98]: Remove default values

```suggestion
set dimension = 2

subsection simulation control
  set output frequency = 0
end

subsection physical properties
  subsection fluid 0
    set kinematic viscosity = 0.01
    set tracer diffusivity  = 1
  end
end

subsection multiphysics
  set tracer = true
end

subsection mesh
  set type               = dealii
  set grid type          = subdivided_hyper_rectangle
  set grid arguments     = 1,  1: -0.1, -0.1: 0.1, 0.1 : true
  set initial refinement = 3
end

subsection boundary conditions
  set number         = 4
  set time dependent = true
  subsection bc 0
    set id   = 0
    set type = function
    subsection u
      set Function expression = 1
    end
  end
  subsection bc 1
    set id   = 1
    set type = outlet
  end
  subsection bc 2
    set id   = 2
    set type = noslip
  end
  subsection bc 3
    set id   = 3
    set type = noslip
  end
end

subsection boundary conditions tracer
  set number = 1
  subsection bc 0
    set id   = 0
    set type = dirichlet
    subsection dirichlet
      set Function expression = 1
    end
  end
end

subsection post-processing
  set verbosity                   = verbose
  set calculate tracer statistics = true
  set tracer statistics name      = tracer_statistics
  set calculate tracer flow rate  = true
  set tracer flow rate name       = tracer_flow_rate_test
end

subsection non-linear solver
  subsection fluid dynamics
    set verbosity      = quiet
    set tolerance      = 1e-4
    set max iterations = 10
  end
  subsection tracer
    set verbosity      = quiet
    set tolerance      = 1e-7
    set max iterations = 30
  end
end

subsection linear solver
  subsection fluid dynamics
    set verbosity = quiet
  end
  subsection tracer
    set verbosity = quiet
  end
end
```

File: source/solvers/tracer.cc [hunk 3]
```diff
@@ -530,6 +553,182 @@ Tracer<dim>::calculate_tracer_statistics()
...
+                               normal_vector_tracer +
+                             tracer_values[q] * velocity_values[q] *
+                               normal_vector_tracer) *
+                            fe_face_values_tracer.JxW(q);
+                        } // end loop on quadrature points
+                    }     // end condition face at boundary
+                }         // end loop on faces
+            }             // end face is a boundary face
+        }                 // end condition cell at boundary
+    }                     // end loop on cells
```
Comment (AmishgaAlphonius) [inline source/solvers/tracer.cc:680]: Why do you check the boundary_ids if you don't select a specific one?

```suggestion
                        } // end loop on quadrature points
                    }     // end condition face at boundary 
                }         // end condition face is a boundary face
            }             // end loop on faces
        }                 // end condition cell at boundary
    }                     // end loop on cells
```

File: source/solvers/tracer.cc [hunk 4]
```diff
@@ -530,6 +553,182 @@ Tracer<dim>::calculate_tracer_statistics()
...
+
+
+  // Sum across all cores
+  for (unsigned int i_bc = 0;
+       i_bc < this->simulation_parameters.boundary_conditions.size;
+       ++i_bc)
+    tracer_flow_rate_vector[i_bc] =
+      Utilities::MPI::sum(tracer_flow_rate_vector[i_bc], mpi_communicator);
+
+  // Filling table
```
Comment (AmishgaAlphonius) [inline source/solvers/tracer.cc:690]:
```suggestion
  // Fill table
```

File: source/solvers/tracer.cc [hunk 5]
```diff
@@ -530,6 +553,182 @@ Tracer<dim>::calculate_tracer_statistics()
...
+                                             tracer_flow_rate_vector[i_bc]);
+      this->tracer_flow_rate_table.set_scientific(
+        "flow-rate-" + Utilities::int_to_string(i_bc, 2), true);
+    }
+
+  // Console output
+  if (simulation_parameters.post_processing.verbosity ==
+      Parameters::Verbosity::verbose)
+    {
+      this->pcout << "Tracer flow rate at the boundaries : " << std::endl;
```
Comment (AmishgaAlphonius) [inline source/solvers/tracer.cc:710]:
```suggestion
      this->pcout << "Tracer flow rate at the boundaries: " << std::endl;
```

Comment (AmishgaAlphonius) [inline source/solvers/tracer.cc:710]: This will impact outputs (you'll need to update your test)

File: source/solvers/tracer.cc [hunk 6]
```diff
@@ -530,6 +553,182 @@ Tracer<dim>::calculate_tracer_statistics()
...
+      for (unsigned int i_bc = 0;
+           i_bc < this->simulation_parameters.boundary_conditions.size;
+           ++i_bc)
+        this->pcout << "\t boundary " << i_bc << " : "
```
Comment (AmishgaAlphonius) [inline source/solvers/tracer.cc:714]:
```suggestion
        this->pcout << "\t boundary " << i_bc << ": "
```

**Review by AmishgaAlphonius — COMMENTED** [review]
Good work! I have a few comments :)

File: source/fem-dem/cfd_dem_coupling.cc [hunk 7]
```diff
@@ -1655,24 +1655,11 @@ CFDDEMSolver<dim>::solve()
   while (this->simulation_control->integrate())
     {
       this->simulation_control->print_progression(this->pcout);
-      bool refinement_step;
-      if (this->simulation_parameters.mesh_adaptation.refinement_at_frequency)
```
Comment (OresteMarquis) [inline source/fem-dem/cfd_dem_coupling.cc:1659]: Just curious to know why is those if conditions are removed from the solvers and fem-dem?

File: source/solvers/tracer.cc [hunk 8]
```diff
@@ -530,6 +553,179 @@ Tracer<dim>::calculate_tracer_statistics()
...
+                  fe_face_values_tracer.reinit(cell, face);
+                  fe_face_values_tracer.get_function_values(
+                    this->present_solution, tracer_values);
+                  fe_face_values_tracer.get_function_gradients(
+                    this->present_solution, tracer_gradient);
+
+                  // We update the fields required by the diffusivity
+                  // model
+                  fields.clear();
+                  if (diffusivity_model->depends_on(field::levelset))
```
Comment (OresteMarquis) [inline source/solvers/tracer.cc:629]: Is there a case where fields don't depend on levelset?

File: source/solvers/tracer.cc [hunk 9]
```diff
@@ -530,6 +553,179 @@ Tracer<dim>::calculate_tracer_statistics()
...
+                      set_field_vector(field::levelset,
+                                       levelset_values,
+                                       fields);
+                    }
+
+                  diffusivity_model->vector_value(fields, tracer_diffusivity);
+
+                  // Get fluid dynamics active cell iterator
+                  typename DoFHandler<dim>::active_cell_iterator cell_fd(
+                    &(*(this->triangulation)),
```
Comment (OresteMarquis) [inline source/solvers/tracer.cc:651]: I'm not a pro in c++ but if feel this is redundant? Doesn't `this -> triangulation` already return the address?

**Review by OresteMarquis — APPROVED** [review]
Really nice! Nothing much to say, only clarifications for my understanding.

File: CHANGELOG.md [hunk 10]
```diff
@@ -5,6 +5,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/).
 
 ## [Master] - 2024-08-05
 
+### Added
+
+- MINOR A post-processing feature for the tracer flow rate through boundaries is added. This allows to produce data useful for residence time distribution analyses. [#1197](https://github.com/chaos-polymtl/lethe/pull/1197)
```
Comment (hepap) [inline CHANGELOG.md:10]: I think you should add the fix on the bc update

File: source/solvers/tracer.cc (hunk 8 above)
Comment (oguevremont) [reply source/solvers/tracer.cc:629]: Yes there can be, for example when we use the tracer without an immersed solid

**Review by hepap — APPROVED** [review]
Nice work! Sorry for the delay in my review!

I think your application test is enough, but a unit test could be also nice :)

File: source/solvers/tracer.cc (hunk 9 above)
Comment (oguevremont) [reply source/solvers/tracer.cc:651]: It's to get the more general type Triangulation (for the function) instead of the parallel::DistributedTriangulationBase that we have

File: source/solvers/tracer.cc (hunk 8 above)
Comment (oguevremont) [reply source/solvers/tracer.cc:629]: Yes that's it, since it's a map there is simply no entry

File: source/solvers/tracer.cc (hunk 9 above)
Comment (OresteMarquis) [reply source/solvers/tracer.cc:651]: Interesting!

File: source/fem-dem/cfd_dem_coupling.cc (hunk 7 above)
Comment (oguevremont) [reply source/fem-dem/cfd_dem_coupling.cc:1377]: It's because in the `update_boundary_conditions` functions, the only condition to respect was whether boundary conditions are time dependent. This is already very restrictive and not expensive to check, so to simplify the code I remove the `refinement_step` boolean here. It had no effect on any test. I feel like the previous condition was difficult to understand and didn't bring much.

File: include/solvers/tracer.h [hunk 11]
```diff
@@ -366,6 +369,20 @@ class Tracer : public AuxiliaryPhysics<dim, GlobalVectorType>
...
   calculate_tracer_statistics();
 
+
+  /**
+   * Post-processing. Calculate the tracer flow rate at every boundary.
+   *
+   * @param current_solution_fd current solution for the fluid dynamics, parsed
+   * by postprocess
+   *
+   * @tparam GlobalVectorType The type of the global vector used for the tracer physics
```
Comment (AmishgaAlphonius) [inline include/solvers/tracer.h:379]:
```suggestion
   * @tparam GlobalVectorType The type of the global vector used for the tracer physic
```

File: source/fem-dem/cfd_dem_coupling.cc (hunk 7 above)
Comment (OresteMarquis) [reply source/fem-dem/cfd_dem_coupling.cc:1377]: If it does the same I like it! It is indeed much clearer :)

**Review by AmishgaAlphonius — APPROVED** [review]
one small correction left 
Nice work! :)

File: source/solvers/tracer.cc (hunk 9 above)
Comment (blaisb) [reply source/solvers/tracer.cc:651]: this->triangulation returns a std::shared_ptr. If you want to have the raw pointer you can do a &(*) trick to obtain the raw pointer behind :)

File: source/solvers/tracer.cc [hunk 12]
```diff
@@ -530,6 +553,179 @@ Tracer<dim>::calculate_tracer_statistics()
...
+       0) &&
+      Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+    {
+      std::string filename =
+        simulation_parameters.simulation_control.output_folder +
+        simulation_parameters.post_processing.tracer_flow_rate_output_name +
+        ".dat";
+      std::ofstream output(filename.c_str());
+      this->tracer_flow_rate_table.write_text(output);
+    }
```
Comment (blaisb) [inline source/solvers/tracer.cc:726]: Personally I would fill the table write the output outside of the function and not inside. The function calculates the flow rate, it does not really take care of anything else. This should be two functions. One two calculate and one to store the information.

**Review by blaisb — CHANGES_REQUESTED** [review]
One small refactoring left and this should be good to go.

**Review by blaisb — APPROVED** [review]

**Comment by blaisb** [comment]
@oguevremont can you rebase and then I'll be able to merge
